Basic setup for the entities module.

louieQ opened the pull request:
@RomitoVincenzo I was thinking about an OO modeling for the units of our analysis (i.e., notebooks and repositories).

Accordingly, each function of our library would be called on `Notebook` objects and take the most suitable representation of notebooks from such objects.

What do you think about it?

louieQ commented:
(the implementation of the proposed refactoring is tentative and incomplete: it should not be merged to master at this point)

louieQ commented on NotebookAnalyzers/entities.py:
```diff
@@ -11,14 +11,20 @@ class Notebook:
         with open(notebook_path.replace(".ipynb", ".py")) as f:
             self.script = f.read()
 
+    def get_code(self):
+        return self.script
+
+    def get_dict(self):
+        return self.nb_dict
```
I'd suggest you to read the first two answers to [this SO post](https://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters). For simple cases like this one, the Pythonic way to use getters and setters is not to use them at all... I know that it's quite counterintuitive, especially for those used to other languages (as I myself was).

Attributes of Python objects are not private (even when their name starts with an underscore `_`): Python does not prevent their access by any means.

From outer scopes, you can directly access the attributes of an object; they should be turned into _properties_ - having getter and setter methods - only if the related get and set operations need to do some preprocessing/postprocessing stuff.

louieQ commented on NotebookAnalyzers/entities.py:
```diff
@@ -11,14 +11,20 @@ class Notebook:
     def __init__(self, notebook_path: str):
         # Importing the notebook as a Python dictionary
-        with open(notebook_path) as f:
+        with open(config.data_path + notebook_path) as f:
```
I believe that the input argument `notebook_path` should contain the full notebook path, rather than its name or a partial path. 

In the future, we might end up having more than one base path for target notebooks. E.g., we might choose to save repos to a different base path. In such cases, handling paths should not be a responsibility of these entity classes.

louieQ commented on NotebookAnalyzers/pynblint.py:
```diff
@@ -4,43 +4,43 @@
-    script = f.read()
-    f.close()
-    return script
+#def notebook_to_dict(filename):
+#    """
+#       Turns a notebook into a dictionary object
+#
+#       Args:
+#           filename(str): name of the notebook file in the data folder
+#       Returns:
+#           data: dictionary object representing the notebook
+#
+#       A way you might use me is
+#
+#       nb_dict = notebook_to_dict("file.ipynb")
+#    """
+#    f = open(config.data_path + filename, )
+#    data = json.load(f)
+#    f.close()
+#    return data
+
+
+#def notebook_to_script(filename):
+#    """
+#       Extracts the code from a jupyter notebook in the data folder
+
+#       Args:
+#           filename(str): name of the notebook file in the TargetNotebook folder
+#       Returns:
+#           script: string containing the python code from the jupyter notebook
+#
+#       A way you might use me is
+#
+#       script = notebook_to_script("file.ipynb")
+#    """
+#    transform_notebook(ipynb_file=config.data_path + filename, export_list=["py"])
+#    f = open(filename.replace(".ipynb", ".py"), 'r')
+#    script = f.read()
+#    f.close()
+#    return script
```
Once we consolidate the new entity classes, these comments can be deleted. 

Should these lines turn to be useful again in the future, we can always checkout one of the old commits and get them back.

louieQ commented on NotebookAnalyzers/NotebooksAnalysisResults.ipynb:
```diff
@@ -53,8 +55,9 @@
-    "              script = pynblint.notebook_to_script(filename)\n",
-    "              nb_dict = pynblint.notebook_to_dict(filename)\n",
+    "              notebook = entities.Notebook(filename)\n",
+    "              script = notebook.get_code()\n",
+    "              nb_dict = notebook.get_dict()\n",
```
I'd rather pass whole `Notebook` objects to the functions of the `pynblint` library.

It should be a responsibility of each function to take the notebook representation it needs from input `Notebook` objects.

RomitoVincenzo commented:
I succesfully refactored the pynblint library, now the *public* functions accept the notebook object. The Notebook __init__ now accepts the whole notebook path and it is therefore more general. Deleted the getters and setters and adapted and added properties to handle dictionary and code representation of the notebook

louieQ commented on NotebookAnalyzers/entities.py:
```diff
@@ -0,0 +1,71 @@
+        with open(notebook_path.replace(".ipynb", ".py")) as f:
+            self._script = f.read()
+
+    @property
+    def nb_dict(self):
+        return self._nb_dict
+
+    @property
+    def script(self):
+        return self._script
+
+    @nb_dict.setter
+    def nb_dict(self, value):
+        self._nb_dict = value
+
+    @script.setter
+    def script(self, value):
+        self._script = value
```
`nb_dict` and `script` can be simple class attributes.

louieQ commented on NotebookAnalyzers/entities.py:
```diff
@@ -48,8 +73,6 @@ class GitHubRepository(Repository):
-        # When it finds a notebook, it appends it to the list of notebooks
-        # e.g., notebooks.append(Notebook(path_where_the_nb_was_found))
-        ...
+        git.Git(config.data_path).clone(github_url)
```
I think a good idea here would be limiting the clone operation to just the last version of the repo.

It should be as easy as adding the `depth=1`  argument to the function call; [here](https://github.com/gitpython-developers/GitPython/issues/269#issuecomment-77743473) is an example.

RomitoVincenzo commented on NotebookAnalyzers/entities.py:
```diff
@@ -28,18 +34,37 @@ class Repository:
+                if file.endswith(".ipynb"):
+                    file_path = os.path.join(root, file).replace('\\', '/')
+                    nb = Notebook(file_path)
+                    nb.path = file_path[8:]
```
As we discussed in the meeting this will not be part of the Notebook constructor.

RomitoVincenzo commented on NotebookAnalyzers/entities.py:
```diff
@@ -28,18 +34,37 @@ class Repository:
+                    nb = Notebook(file_path)
+                    nb.path = file_path[8:]
+                    self.notebooks.append(nb)
+                    os.remove(file[:-5] + "py")
```
As we discussed in the meeting this will not be part of the Notebook constructor.

RomitoVincenzo commented on NotebookAnalyzers/entities.py:
```diff
@@ -28,18 +34,37 @@ class Repository:
+        for root, dirs, files in os.walk(repo_path):
+            for file in files:
+                if file.endswith(".ipynb"):
+                    file_path = os.path.join(root, file).replace('\\', '/')
```
This is the only task i could not manage to solve. I tried to use ' pathlib.Path.joinpath(root,file) ' at line 41 insted of the os.path.join but it raises some kind of error i can't understand the meaning.

RomitoVincenzo commented on NotebookAnalyzers/entities.py:
```diff
@@ -23,9 +24,29 @@ def __init__(self, notebook_path: str):
             self.script = f.read()
 
+    def get_info(self, bottom_size=4):
+        notebook_tuple = (self.path,
```
@louieQ i write here the same I told you on Teams just to keep the pull updated. the function now returns a dict object containing all the the pynblint results. Only thing is that I could not mantain the structure of the API dict response with the nested dicts (statistics,linting) because I could not represent them in the dataframe(can't find a way to represent a list of nested dictionaries).

louieQ commented:
@RomitoVincenzo I've made a bit of refactoring on this branch, which now I think is ready to be merged.

Please, test whether it works on your machine too. You can use the example notebook, which I modified to reflect the changes in the library.

Apart from making a consistent use of `pathlib.Path` to handle project paths, the biggest change I made relates to how notebooks are imported. I dropped the `notebooktoall` dependency and relied on `nbconvert` for the conversion of notebooks to Python scripts. By doing this, I avoided the I/O operations required by `notebooktoall`: by now, once the notebook have been read once, all the computation happens in memory.

louieQ commented on NotebookAnalyzers/main.py:
```diff
@@ -23,39 +25,12 @@ def index():
-    return response
+    nb = Notebook.from_string(notebook.filename)
+    os.remove(Path(config.data_path) / notebook.filename)
+    #os.remove(notebook.filename[:-5] + "py") useless if we use notebook.from_string method
```
Once you delete this comment, feel free to merge

RomitoVincenzo commented on NotebookAnalyzers/entities.py:
```diff
@@ -0,0 +1,99 @@
+        # Importing the notebook as a Python script
+        transform_notebook(ipynb_file=notebook_path, export_list=["py"])
+        notebook_path = notebook_path.split("/")[-1]
+        self.path = notebook_path
```
In this case yes, it is. But when the API calls the end-node to lint a repo or a local project, self.path will contain the notebook's path relative to the project's root folder.
